subsys/net/lib/lwm2m/lwm2m_rw_senml_cbor.c

```diff
@@ -0,0 +1,901 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME net_lwm2m_senml_cbor
+#define LOG_LEVEL CONFIG_LWM2M_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <stdio.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <ctype.h>
+#include <sys/util.h>
+
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+
+#include "lwm2m_engine.h"
+#include "lwm2m_object.h"
+#include "lwm2m_rw_senml_cbor.h"
+#include "lwm2m_senml_cbor_decode.h"
+#include "lwm2m_senml_cbor_encode.h"
+#include "lwm2m_senml_cbor_types.h"
+#include "lwm2m_util.h"
+
+struct cbor_out_formatter_data {
+	/* Data */
+	struct lwm2m_senml input;
+
+	/* Storage for basenames and names ~ sizeof("/65535/999/") */
+	struct {
+		char names[CONFIG_LWM2M_RW_SENML_CBOR_RECORDS][sizeof("/65535/999/")];
+		size_t name_sz; /* Name buff size */
+		uint8_t name_cnt;
+	};
+};
+
+struct cbor_in_fmt_data {
+	/* Decoded data */
+	struct lwm2m_senml dcd; /* Decoded data */
+	struct record *current;
+	char basename[sizeof("/65535/999/")]; /* Null terminated basename */
+};
+
+#define GET_CBOR_FD_NAME(fd) ((fd)->names[(fd)->name_cnt])
+/* Get the current record */
+#define GET_CBOR_FD_REC(fd) \
+	&((fd)->input._lwm2m_senml__record[(fd)->input._lwm2m_senml__record_count])
+/* Get a record */
+#define GET_IN_FD_REC_I(fd, i) &((fd)->dcd._lwm2m_senml__record[i])
+/* Consume the current record */
+#define CONSUME_CBOR_FD_REC(fd) \
+	&((fd)->input._lwm2m_senml__record[(fd)->input._lwm2m_senml__record_count++])
+/* Get CBOR output formatter data */
+#define LWM2M_OFD_CBOR(octx) ((struct cbor_out_formatter_data *)engine_get_out_user_data(octx))
+
+static struct cbor_out_formatter_data *setup_out_fmt_data(struct lwm2m_message *msg)
+{
+	struct cbor_out_formatter_data *fd = malloc(sizeof(*fd));
```

Comment:
Do we really need to involve heap here? Couldn't senml cbor do the same as json, i. e. allocate the context on a stack?

Reply:
I tried with it and ended up needing 14K of stack space... If someone finds a way to reduce the memory consumption then using stack might be an option. I don't have any ideas about how to do it at the moment though.

Reply:
What's so large, `struct lwm2m_senml`? Anyway, maybe a better option then would be to have a single preallocated in/out context and protect it with a mutex? I think we could even keep the contexts in a union to save space. The encoder will mostly be used from a single thread anyway (well, with exception to LwM2M send perhaps). 

I'd really would like to prevent a situation where the SenML cbor encoder is rendered useless due to some other component hogging heap in the system.

Reply:
`struct lwm2m_senml`, yeah, that's the one.

Reply:
@oyvindronningstad Any good ideas on how to optimize the code?

Reply:
> The encoder will mostly be used from a single thread anyway (well, with exception to LwM2M send perhaps).
Would it make sense if a SEND operation would be executed by using the same queue?

Reply:
I'm not sure I follow, the encoding in `lwm2m_engine_send()` takes place in the caller thread, I don't think there's an easy way to delegate this operation to LwM2M thread. Once encoded, the LwM2M message is queued and processed by the LwM2M thread (i. e. sent over a socket).

Reply:
And just to clarify - I don't think that it's a big deal that encoding can take place from a different thread. If we implment a proper mutex protection for the shared senml cbor context, the only effect would be a slight delay in one of the threads.

Reply:
> Any good ideas on how to optimize the code?

From my rough count, `struct lwm2m_senml` should take about 3k of space (30 * ~100). You can likely save a few bytes on each record by not using `numeric`:

Change
```
value = tstr / bstr / numeric / bool
```
to
```
value = tstr / bstr / int .size 8 / float / bool
```

You can also try changing
```
	? ( v => numeric  //   ; Numeric Value
```
to
```
	? ( v => int .size 8  //   ; Numeric Value
		v => float //   ; Numeric Value
```
Though that might increase code size.

Reply:
> You can also try changing
> 
> ```
> 	? ( v => numeric  //   ; Numeric Value
> ```
> 
> to
> 
> ```
> 	? ( v => int .size 8  //   ; Numeric Value
> 		v => float //   ; Numeric Value
> ```

That doesn't seem to work
```
zcbor --default-max-qty 99 -c senml_stripped_v2.cddl code -d -t lwm2m_senml --oc lwm2m_senml_cbor_decode2.c --oh lwm2m_senml_cbor_decode2.h
Parsing files: senml_stripped_v2.cddl, /home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/cddl/prelude.cddl
Traceback (most recent call last):
  File "/home/vep1/.virtualenvs/zephyr/bin/zcbor", line 8, in <module>
    sys.exit(main())
  File "/home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/zcbor.py", line 2960, in main
    args.process(args)
  File "/home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/zcbor.py", line 2884, in process_code
    renderer = CodeRenderer(entry_types={mode: [cddl_res[mode].my_types[entry]
  File "/home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/zcbor.py", line 2388, in __init__
    self.functions[mode] = self.unique_funcs(mode)
  File "/home/vep1/.virtualenvs/zephyr/lib/python3.8/site-packages/zcbor/zcbor.py", line 2433, in unique_funcs
    assert func_names[func_name][0] == func_xcode,\
AssertionError: Two elements share the function name decode_union_v, but their implementations are not identical. Please change one or both names.

(((zcbor_uint32_expect(state, (2))))
	&& (zcbor_int64_decode(state, (&(*result)._union_v)))
	&& ((((*result)._union_v >= -9223372036854775807LL)
	&& ((*result)._union_v <= 9223372036854775807LL)) || (zcbor_error(state, ZCBOR_ERR_WRONG_RANGE), false)))

(((zcbor_uint32_expect(state, (2))))
	&& (zcbor_float_decode(state, (&(*result)._union_v))))
[ncs2/zephyr|zephyr|]:~/ws$ cat senml_stripped_v2.cddl 
lwm2m_senml = [1* record]

record = {
	? bn => tstr,           ; Base Name
	? n => tstr,            ; Name
	? ( v => int .size 8 // ; Numeric Value
	    v => float       // ; Numeric Value
	    vs => tstr       // ; String Value
	    vb => bool       // ; Boolean Value
	    vd => bstr ),       ; Data Value
	0*3 key-value-pair      ; To handle unordered maps; length-first ordered map keys
}

; now define the generic versions
key-value-pair = ( int => value )

value = tstr / bstr / int .size 8 / float / bool

n  = 0
bn = -2
v  = 2
vs = 3
vd = 8
vb = 4
````

Reply:
This seems to work (using two different aliases for `v`):

```
lwm2m_senml = [1* record]

record = {
	? bn => tstr,           ; Base Name
	? n => tstr,            ; Name
	? ( vi => int .size 8 // ; Numeric Value
	    vf => float       // ; Numeric Value
	    vs => tstr       // ; String Value
	    vb => bool       // ; Boolean Value
	    vd => bstr ),       ; Data Value
	0*3 key-value-pair      ; To handle unordered maps; length-first ordered map keys
}

; now define the generic versions
key-value-pair = ( int => value )

value = tstr / bstr / int .size 8 / float / bool

n  = 0
bn = -2
vi  = 2
vf  = 2
vs = 3
vd = 8
vb = 4
```

Reply:
Another thing that I have been pondering is that instead of having int64s should I turn those into int32s and uint32s. The mandatory objects won't need 64-bit values but do we need to be prepared that someone would introduce an object which would use those?

Reply:
@oyvindronningstad I regenerated the code based on your suggestions and made the corresponding adjustments.

Reply:
>instead of having int64s should I turn those into int32s and uint32s

@VeijoPesonen This would not save you any space, since they are in unions with tstr/bstr/float which use 8 bytes.

Reply:
> What's so large, `struct lwm2m_senml`? Anyway, maybe a better option then would be to have a single preallocated in/out context and protect it with a mutex? I think we could even keep the contexts in a union to save space. The encoder will mostly be used from a single thread anyway (well, with exception to LwM2M send perhaps).

@rlubos Should be done now as you requested.